src/topology/algebra/uniform_ring.lean

```diff
@@ -150,8 +156,66 @@ instance : comm_ring (completion R) :=
       (assume a b, by rw [← coe_mul, ← coe_mul, mul_comm]),
  ..completion.ring }
 
-end uniform_space.completion
+@[simp] lemma map_mul_eq_mul_coe (r : R) :
+  completion.map ((*) r) = (*) (r : completion R) :=
+begin
+  ext x,
+  refine completion.induction_on x _ (λ s, _),
+  { exact is_closed_eq (completion.continuous_map) (continuous_mul_left (r : completion R)) },
+  { rw [map_coe (uniform_continuous_mul_const R r), coe_mul] },
+end
+
+instance : algebra R (completion R) :=
+{ commutes' := λ r x, by simp only [ring_hom.to_fun_eq_coe, mul_comm],
+  smul_def' := λ r x, congr_fun (map_mul_eq_mul_coe R r) x,
+  .. (uniform_space.completion.coe_ring_hom : R →+* completion R) }
 
+lemma algebra_smul_eq (r : R) (x : (completion R)) :
+  r • x = (r : completion R) * x :=
+by rw algebra.smul_def; refl
+
+section algebra
+variables (S : Type*) [comm_semiring S] [algebra S R] [has_uniform_continuous_const_smul S R]
+
+@[simp] lemma map_smul_eq_mul_coe (s : S) :
+  completion.map ((•) s) = (*) (algebra_map S R s : completion R) :=
+begin
+  ext x,
+  refine completion.induction_on x _ (λ r, _),
+  { exact is_closed_eq (completion.continuous_map) (continuous_mul_left _) },
+  { rw [map_coe (has_uniform_continuous_const_smul.uniform_continuous_const_smul s),
+      algebra.smul_def, coe_mul];
+    apply_instance, },
+end
+
+instance algebra' : algebra S (completion R) :=
+{ commutes' := λ s x, by simp only [ring_hom.to_fun_eq_coe, mul_comm],
+  smul_def' := λ s x, congr_fun (map_smul_eq_mul_coe R S s) x,
+  ..((uniform_space.completion.coe_ring_hom : R →+* completion R).comp (algebra_map S R)) }
+
+@[simp] lemma algebra'_smul_eq (s : S) (x : completion R) :
+  s • x = (algebra_map S R s : completion R) * x :=
+by rw algebra.smul_def; refl
+
+lemma coe_algebra_map (s : S) :
+  (algebra_map S R s : completion R) = algebra_map S (completion R) s :=
+rfl
+
+instance is_scalar_tower {T : Type*} [comm_semiring T] [algebra S T] [algebra T R]
+  [is_scalar_tower S T R] [has_uniform_continuous_const_smul T R] :
+  is_scalar_tower S T (completion R) :=
+⟨λ r k x,
+begin
+  simp only [← mul_assoc, algebra.smul_def, map_mul],
+  congr,
+  rw [← coe_algebra_map, ← coe_algebra_map],
+  simp only [algebra.algebra_map_eq_smul_one],
+  rw [smul_assoc, one_smul],
+end⟩
```

Comment:
I suspect this is still less general than it can be. The version I'm thinking of is
```lean
instance [has_scalar N X] [has_scalar M N]
  [has_uniform_continuous_const_smul M X] [has_uniform_continuous_const_smul N X]
  [is_scalar_tower M N X] : is_scalar_tower M N (completion X) :=
⟨λ m n x, begin
  have : _ = (_ : completion X → completion X) :=
    map_comp (uniform_continuous_const_smul m) (uniform_continuous_const_smul n),
  refine eq.trans _ (congr_fun this.symm x),
  exact congr_arg (λ f, completion.map f x) (by exact funext (smul_assoc _ _)),
end⟩
```
which goes below https://github.com/leanprover-community/mathlib/blob/3dbb894f9623bb03b0f1cb2a60f9c54d7a0cfc68/src/topology/algebra/uniform_mul_action.lean#L95-L99

You should be able to write an `smul_comm_class` instance in the same way.